File: mirrorTree.java
// Time Complexity :O(n)
// Space Complexity :O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
class Solution {
    public boolean isSymmetric(TreeNode root) {
        
        if(root==null){
            return true;
        }
        return recurr(root.left,root.right);
        
    }
    private boolean recurr(TreeNode p1, TreeNode p2){
        
        if(p1==null && p2==null){
            return true;
        }
        
        if(p1==null || p2==null || p1.val!=p2.val){
            return false;
        }
        
        return recurr(p1.left, p2.right) && recurr(p1.right, p2.left);
    }
}

File: Mirror_Tree.cpp

 //dfs Solution
class Solution {
    bool flag;
public:
    bool isSymmetric(TreeNode* root) {
        flag = true;
        dfs(root->left,root->right);
        return flag;
    }
private:
    void dfs(TreeNode* left_child, TreeNode* right_child){
     //base
        if(left_child==NULL && right_child==NULL)return;
     //logic
        if(left_child==NULL || right_child==NULL ||                     left_child->val!=right_child->val){             
            flag=false;
            return;
        }
        dfs(left_child->left, right_child->right);
        dfs(left_child->right, right_child->left);
    }    
};
/*
Conditional Recursion

 //dfs Solution
class Solution {
    bool flag;
public:
    bool isSymmetric(TreeNode* root) {
        flag = true;
        dfs(root->left,root->right);
        return flag;
    }
private:
    void dfs(TreeNode* left_child, TreeNode* right_child){
     //base
        if(left_child==NULL && right_child==NULL)return;
     //logic
        if(left_child==NULL || right_child==NULL ||                     left_child->val!=right_child->val){             
            flag=false;
            return;
        }
        if(flag){
            dfs(left_child->left, right_child->right);
        }
        if(flag){      
        dfs(left_child->right, right_child->left);
        }
    }    
};

*/
